cilium/CFP-23380-ordered-policy.md

```diff
@@ -0,0 +1,329 @@
+    rule2[prio 2, tcp, 80, id *]
+    root[[root]]
+    
+    root-->rule1
+    rule1-. invalid .->rule2
+```
+
+**Example 1c:** Can we insert `(tcp, port *, identity 5)`?
+    Yes, because this has lower precedence.
+
+```mermaid
+---
+title: "Example 1c: valid"
+---
+graph TD
+    rule1[prio 1, tcp, 0-1023, id *]
+    rule2[prio 2, tcp, *, id 5]
+    root[[root]]
+    
+    root-->rule2
+    rule2-->rule1
+```
+
+**Example 1d:** Can we insert `(udp, port 53, identity 5)`?
+    Yes, because this does not cover a higher-priority entry.
+    
+```mermaid
+---
+title: "Example 1d: valid"
+---
+graph TD
+    rule1[prio 1, tcp, 0-1023, id *]
+    id2[prio 2, udp, *53, id 5]
+    root[[root]]
+    
+    root-->rule1
+    root-->id2
+```
+
+#### Wildcard and non-wildcard interaction
+
+Special care must be taken for identity wildcards, both when higher and lower precedence.
+
+If a lower-priority non-wildcard PerSelectorPolicy would override an existing wildcard, higher-priority, lower-precedence selector, adding it is blocked. We must also flag this selector as covered, so we ignore any incremental updates. See Example 2a.
+
+If a lower-priority wildcard PerSelectorPolicy would override an existing non-wildcard, higher-priority, lower-precedence selector, then it must be inserted, but covering entries from the higher-priority selector must be synthesized. See example 2b.
+
+#### Example 2a: Ancestor wildcard
+
+Three rules:
+
+1. `(tcp, port *, id {*})`
+2. `(tcp, port 80, id {*})`
+3. `(tcp, port 80, id {5})`
+
+Rules 2 and 3 would override rule 1, so they may not be inserted. Rule 3 must also be flagged as disabled, so that incremental updates are not inadvertently added.
+
+```mermaid
+
+graph TB
+    rule1["prio 1, tcp, 0-65535, id {*}"]
+    rule2["prio 2, tcp, 80, id {*}, disabled"]
+    rule3["prio 3, tcp, 80, id {5}, disabled"]
+    root[[root]]
+    
+    root --> rule1
+    rule1 -. invalid .-> rule2 & rule3
+```
+
+#### Example 2b: Descendant wildcard
+
+Two rules:
+1. `(tcp, port *, id {5, 6})`
+2. `(tcp, port 80, id {*})`
+
+Rule 2 is higher precedence. For ID 5, rule 2 overrides rule 1, which is an error. However, rule 2's wildcard selector selects more identities, so it cannot be elided. Rather, we explicitly insert covering entry `(tcp, port 80, id {5, 6})`. Likewise, we flag rule 1 as "covered" by rule 2, so we know to manage incremental updates.
+
+```mermaid
+
+graph TB
+    rule1["prio 1, tcp, 0-65535, id {5, 6}"]
+    rule2["prio 2, tcp, 80, id {*}"]
+    rule1a["prio 1, tcp, 80, id {5, 6}"]
+    root[[root]]
+    
+    root --> rule1
+    rule1 --> rule2 & rule1a
+    rule2 -. covers .-> rule1
+```
+
+#### Merging mutual authentication
+
+Whether or not a connection requires mututal authentication is *also* managed by the network policy engine. However, authentication is not a "traffic verdict". (Traffic verdicts are `allow`, `deny`, and `proxy`). Rather, it is applied to all traffic, regardless of selectors.
```

Comment:
nit: I won't go over them all but a spellchecker could pick up on some things like this. I also noticed some cases like `PerSecectorPolicies` that a spellchecker won't pick up on by default but would also be nice to fix.
```suggestion
Whether or not a connection requires mutual authentication is *also* managed by the network policy engine. However, authentication is not a "traffic verdict". (Traffic verdicts are `allow`, `deny`, and `proxy`). Rather, it is applied to all traffic, regardless of selectors.
```